scrunch/order.py

```diff
@@ -0,0 +1,576 @@
+            self.order.update()
+            return
+        elements = self._validate_alias_arg(alias)
+        position = 0 if (before or after) else position
+
+        # create the new Group obj and insert all `elements`
+        new_group = Group({name: []}, order=self.order, parent=self)
+        new_group.insert(elements)
+
+        # add the new Group to self.elements so that `insert` detects it
+        self.elements[name] = new_group
+        self.insert(new_group.name, position=position,
+                    before=before, after=after)
+
+    def rename(self, name):
+        name = self._validate_name_arg(name)
+
+        if self.is_root:
+            raise ValueError(
+                'Renaming the root Group is not allowed.'
+            )
+
+        if name in self.parent.elements:
+            raise ValueError(
+                'Parent Group \'%s\' already contains an element named \'%s\'.'
+                % (self.parent.name, name)
+            )
+
+        if name == self.name:
+            # Nothing to do.
+            return
+
+        # Rename!
+        _elements = collections.OrderedDict()
+        for current_name, obj in self.parent.elements.items():
+            if current_name == self.name:
+                _elements[name] = obj
+            else:
+                _elements[current_name] = obj
+        self.name = name
+        self.parent.elements = _elements
+
+        # Update!
+        self.order.update()
+
+    def move(self, path, position=-1, before=None, after=None):
+        position = 0 if (before or after) else position
+        path = Path(path)
+        if not path.is_absolute:
+            raise InvalidPathError(
+                'Invalid path %s: only absolute paths are allowed.' % path
+            )
+        target_group = self.order[str(path)]
+
+        if target_group == self:
+            raise InvalidPathError(
+                'Invalid path %s: cannot move Group into itself.' % path
+            )
+        target_group.insert(
+            self.name, position=position, before=before, after=after
+        )
+
+
+class Order(object):
+
+    def __init__(self, resource):
+        self.resource = resource
+        self._hier = None
+        self._graph = None
+        self._sync = True
+
+    def _load_hier(self):
+        """
+        subclasses need to overwrite this method in order to get the updated
+        order hierarchy
+        """
+        return self._hier
+
+    def _load_graph(self):
+        self._graph = Group({'__root__': self.hier.graph}, order=self)
+        return self._graph
+
+    @property
+    def hier(self):
+        if self._hier is None:
+            self._load_hier()
+        return self._hier
+
+    @hier.setter
+    def hier(self, _):
+        raise TypeError('Unsupported assignment operation')
+
+    @property
+    def graph(self):
+        if self._graph is None:
+            self._load_graph()
+        return self._graph
+
+    @graph.setter
+    def graph(self, _):
```

Comment:
I think this is overkill, if we just set the `self.group = Group(...)` to be declared on init it will be a normal attribute, we cannot be protecting all attributes in python from being replaced, if a user is setting internal library attributes to something else, it will be their problem fro doing wrong usage. 

`self.order` and `self.group` don't need to be lazy nor to have `load_*` methods.